api/proto/metrics/metrics.proto

```diff
@@ -177,12 +177,39 @@ message appCpuMetric {
   uint64 total_ns = 7;                 // like 'total' but in nanoseconds. Overflow may happen after ~40 years without EVE restart
 }
 
-// StorageMetric   At the moment not used
+// StorageDeviceMetrics - contains device statistics from ZFS storage
+message StorageDeviceMetrics {
+  uint64 alloc = 1;             // space allocated (in byte)
+  uint64 total = 2;             // total capacity (in byte)
+  uint64 deflated_space = 3;    // deflated capacity (in byte)
+  uint64 replaceable_size = 4;  // replaceable dev size (in byte)
+  uint64 expandable_size = 5;   // expandable dev size (in byte)
+  uint64 read_errors = 6;       // counter read count errors
+  uint64 write_errors = 7;      // counter write errors
+  uint64 checksum_errors = 8;  // counter checksum errors
+  uint64 bytes_read = 9;       // bytes read (in byte)
+  uint64 bytes_write = 10;      // bytes written (in byte)
+  uint64 ops_count_read = 11;   // Number of read ops
+  uint64 ops_count_write = 12;  // Number of write ops
+  uint64 ios_in_progress = 13;  // number of I/Os currently in progress.
+  uint64 read_ticks = 14;       // The total number of milliseconds spent by all reads.
+  uint64 write_ticks = 15;      // The total number of milliseconds spent by all writes.
+  uint64 ios_total_ticks = 16;  // The number of milliseconds spent doing I/Os.
+  // weighted_io_ticks is the weighted number of milliseconds spent doing I/Os.
+  // This can also be used to estimate average queue wait time for requests.
+  uint64 weighted_io_ticks = 17;
+}
+
+// StorageMetric - contains zpool statistics including child devices
 message StorageMetric {
-  string pool_name = 1;
-  // Depreated tag 2
-  // Depreated tag 3
-  // Depreated tag 4
+  // Deprecated = 1;
```

Comment:
Why remove the pool name?

Reply:
Since the same message is used and for child elements that are included in the zpool. I decided to add a more generic name "display_name" and added a comment to explain it. I did this because I didn't want to be misleading, since it wouldn't make sense to store the disk or raid name in the zpool_name variable. And storing the pool name for each device also does not make sense. In addition, it is always guaranteed that the first element in this message will be zpool.

Reply:
Don't we have a similar structure for the info messages? I think it makes sense to use the same naming scheme for the top-level object (the pool) and the components/children as for the info message. Can you check what is different and what is the same?

Reply:
Sorry for the delay,
In answer to your question, the structure and hierarchy were identical to the info message, except that the types were explicitly separated into separate messages in the info message.

 I rewrote the structure.
- Created a separate message for each type (Pool/RAID/Disk/Zvol).
- Added GUID for RAID/Mirror info message. Because the DisplayName value is not unique, metric matching may have issues.
- Added metrics for Zvol from /proc/diskstats

Question.
- Can we get by with a single unique value (eg serial_number for disks, GUID for RAID/mirror, etc.) in the metrics messages? I think that if the controller can associate information from metrics with information from info messages, then there is no point in sending additional information to identify metrics.

Reply:
My point is that the info and metrics messages must use the same identifying information for the objects; anything else is just too confusing.
What information is needed to identify a particular object is then the next question. For disks I think we already defined a message which includes the serial number but has other fields as well (partly this is because we might about to allow the user to define some friendly names and also to allow zedcloud to create names which are not reused if a disk is replaced).
Question: what names are we currently using for the ZFS vdevs in the info messages?